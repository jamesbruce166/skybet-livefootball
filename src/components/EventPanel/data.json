{
	"empty-data-message": "There are no games to display at the present time.",
	"menu-odds-label": "Odds Format",
	"format-options": {
		"decimal": "Decimal (1.5)",
		"fraction": "Fraction (1/2)"
	}
}
