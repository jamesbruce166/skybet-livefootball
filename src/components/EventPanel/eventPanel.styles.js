import styled from 'styled-components';
import { PRIMARY_RED } from '../../variables.styles';

export const Section = styled.div`
	background: ${PRIMARY_RED};
	min-height: 20vh;
	padding: 5%;
	position: -webkit-sticky;
	position: sticky;
	top: 0;
`;

export const GameBox = styled.div`
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: stretch;
`;

export const TeamTextContainer = styled.div`
	display: flex;
`;

export const TeamText = styled.p`
	display: flex;
	flex: 1;
	justify-content: center;
	color: white;
	font-size: 22px;
	text-align: center;

	:first-child > span {
		margin-right: auto;
	}

	:last-child > span {
		margin-left: auto;
	}
`;

export const TimeLabel = styled.p`
	color: lightgray;
	font-size: 14px;
	text-align: center;
	text-transform: uppercase;
	line-height: 0;
`;

export const TimeStarted = styled.p`
	color: white;
	font-size: 16px;
	text-align: center;
	line-height: 0;
`;

export const MarketContainer = styled.div`
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: stretch;
	margin-top: 10px;
	transition: 0.3s ease-in-out;
`;

export const Outcome = styled.div`
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	margin-top: 10px;
`;

export const OutcomeName = styled.p`
	color: lightgray;
	font-size: 16px;
	text-align: center;
	text-transform: uppercase;
	line-height: 0;
`;

export const OutcomePrice = styled.p`
	color: white;
	font-size: 14px;
	text-align: center;
`;

export const Tabs = styled.div`
	display: flex;
	flex-direction: row;
	justify-content: center;
	padding: 0 30px;
	overflow: hidden;
	background: transparent;
	height: 3em;
	margin-top: 20px;
`;

export const Tab = styled.button`
	cursor: pointer;
	position: relative;
	background: transparent;
	color: lightgray;

	margin: 0.3em;
	font-size: 0.9em;
	border: none;
	transition: 0.3s ease-in-out;

	${(props) => (props.active ? 'font-size: 1rem;' : '')}
	${(props) => (props.active ? 'margin: 0 0.6rem;' : '')}
    ${(props) => (props.active ? 'color: white;' : '')}
`;

export const Content = styled.div`
	${(props) => (props.active ? 'display: flex;' : 'display:none;')}
	flex-direction: row;
	justify-content: space-around;
	align-items: center;
`;

export const EmptyDataText = styled.p`
	color: white;
	font-size: 20px;
	text-align: center;
	font-weight: 500;
`;
