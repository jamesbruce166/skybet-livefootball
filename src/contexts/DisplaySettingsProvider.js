import React, { useContext, useState } from 'react';

export const DisplaySettingsContext = React.createContext();

export function useDisplaySettings() {
	return useContext(DisplaySettingsContext);
}

export function DisplaySettingsProvider({ children }) {
	const displayOptions = {
		FRACTION: 'decimal',
		DECIMAL: 'fraction',
	};
	const [oddsDisplay, setOddsDisplay] = useState(displayOptions.DECIMAL);

	const value = {
		oddsDisplay,
		setOddsDisplay,
		displayOptions,
	};

	return (
		<DisplaySettingsContext.Provider value={value}>
			{children}
		</DisplaySettingsContext.Provider>
	);
}
