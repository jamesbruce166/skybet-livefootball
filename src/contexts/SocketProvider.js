import React, { useContext, useEffect, useState, useRef } from 'react';

const SocketContext = React.createContext();

export function useSocket() {
	return useContext(SocketContext);
}

export function SocketProvider({ id, children }) {
	const [socket, setSocket] = useState();
	const ws = useRef(null);

	useEffect(() => {
		ws.current = new WebSocket(process.env.REACT_APP_SOCKET_URL);

		ws.current.addEventListener('open', () => {
			ws.current.send(
				JSON.stringify({
					type: 'getLiveEvents',
					primaryMarkets: true,
				})
			);
		});
		setSocket(ws.current);
	}, [id]);

	const value = {
		socket,
	};

	return (
		<SocketContext.Provider value={value}>
			{children}
		</SocketContext.Provider>
	);
}
