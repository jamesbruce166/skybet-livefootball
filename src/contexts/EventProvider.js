import React, {
	useContext,
	useState,
	useCallback,
	createContext,
	useEffect,
} from 'react';
import { useErrorHandler } from 'react-error-boundary';

import { useSocket } from './SocketProvider';

export const EventsContext = createContext();

export function useEvents() {
	return useContext(EventsContext);
}

export function EventProvider({ children }) {
	const [selectedEvent, setSelectedEvent] = useState(undefined);
	const [events, setEvents] = useState(undefined);
	const handleError = useErrorHandler();

	const { socket } = useSocket();

	const onMessage = useCallback((m) => {
		const data = JSON.parse(m.data);
		const { type } = data;
		switch (type) {
			case 'LIVE_EVENTS_DATA':
				const firstGame = data.data && data.data[0];

				setEvents(data);
				setSelectedEvent(firstGame);
				break;
			case 'ERROR':
				handleError(data);
				break;
			default:
				break;
		}
	}, []);

	useEffect(() => {
		if (socket == null) return;

		socket.addEventListener('message', onMessage);

		return () => {
			socket.removeEventListener('message', onMessage);
		};
	}, [socket, onMessage]);

	const value = {
		events,
		selectedEvent,
		setSelectedEvent,
	};

	return (
		<EventsContext.Provider value={value}>
			{children}
		</EventsContext.Provider>
	);
}
