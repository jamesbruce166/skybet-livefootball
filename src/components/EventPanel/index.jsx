import React, { useState } from 'react';
import {
	Menu,
	MenuItem,
	SubMenu,
	MenuHeader,
	MenuRadioGroup,
} from '@szhsin/react-menu';
import '@szhsin/react-menu/dist/index.css';
import {
	Section,
	GameBox,
	TeamTextContainer,
	TeamText,
	TimeLabel,
	TimeStarted,
	EmptyDataText,
	ControlBar,
	MenuIcon,
	MinimiseIcon,
	MaximiseIcon,
	Clickable,
	LinkedEventTypeName,
	ControlBarLeft,
	ControlBarRight,
} from './eventPanel.styles';
import MarketTabs from '../MarketTabs';
import { useEvents } from '../../contexts/EventProvider';
import { useDisplaySettings } from '../../contexts/DisplaySettingsProvider';

import displayData from './data';

const EventPanel = () => {
	const { selectedEvent } = useEvents();
	const { oddsDisplay, setOddsDisplay, displayOptions } =
		useDisplaySettings();
	const [isMinimised, setIsMinimised] = useState(true);

	const minimiseHandler = () => setIsMinimised(!isMinimised);
	const radioChangeHandler = (e) => setOddsDisplay(e.value);

	const EmptyData = () => {
		return (
			<EmptyDataText>{displayData['empty-data-message']}</EmptyDataText>
		);
	};

	const SelectedEventData = () => {
		const { competitors, scores, startTime, eventId } = selectedEvent;

		const { name: homeTeamName } = competitors[0];
		const { name: awayTeamName } = competitors[1];
		const { home, away } = scores;

		const renderStartTime = (d) => {
			const date = new Date(d);
			const hour = date.getHours();
			const minutes = date.getMinutes();
			return `${hour}:${minutes}`;
		};

		return (
			<>
				<GameBox>
					<TeamTextContainer>
						<TeamText>{homeTeamName}</TeamText>
						<TeamText>
							{home} - {away}
						</TeamText>
						<TeamText>{awayTeamName}</TeamText>
					</TeamTextContainer>
					<TimeLabel>Started At</TimeLabel>
					<TimeStarted>{renderStartTime(startTime)}</TimeStarted>
				</GameBox>
				{!isMinimised && (
					<MarketTabs
						displayOptions={displayOptions}
						eventId={eventId}
					/>
				)}
			</>
		);
	};

	return (
		<Section data-testid='selected-event'>
			<ControlBar>
				<ControlBarLeft>
					<LinkedEventTypeName>
						{selectedEvent &&
							(selectedEvent.linkedEventTypeName ??
								'Football Live')}
					</LinkedEventTypeName>
				</ControlBarLeft>
				<ControlBarRight>
					<Menu menuButton={<MenuIcon />}>
						<MenuHeader>Display Settings</MenuHeader>
						<SubMenu label={displayData['menu-odds-label']}>
							<MenuRadioGroup
								value={oddsDisplay}
								onRadioChange={radioChangeHandler}
							>
								<MenuItem
									type='radio'
									value={displayOptions.FRACTION}
								>
									{displayData['format-options'].fraction}
								</MenuItem>
								<MenuItem
									type='radio'
									value={displayOptions.DECIMAL}
								>
									{displayData['format-options'].decimal}
								</MenuItem>
							</MenuRadioGroup>
						</SubMenu>
					</Menu>
					<Clickable onClick={minimiseHandler}>
						{isMinimised ? <MinimiseIcon /> : <MaximiseIcon />}
					</Clickable>
				</ControlBarRight>
			</ControlBar>
			{selectedEvent ? <SelectedEventData /> : <EmptyData />}
		</Section>
	);
};

export default EventPanel;
