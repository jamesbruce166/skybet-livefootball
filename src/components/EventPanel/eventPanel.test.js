import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, cleanup } from '@testing-library/react';

import { EventsContext } from '../../contexts/EventProvider';
import { SocketContext } from '../../contexts/SocketProvider';
import { DisplaySettingsContext } from '../../contexts/DisplaySettingsProvider';

import displayData from './data.json';
import EventPanel from './index';
import mockEvent from './__mocks__/event.json';

afterEach(() => {
	cleanup();
});

function renderEventPanel(event) {
	const displayOptions = {
		FRACTION: 'decimal',
		DECIMAL: 'fraction',
	};
	const oddsDisplay = displayOptions.DECIMAL;
	return (
		<DisplaySettingsContext.Provider
			value={{ oddsDisplay, displayOptions }}
		>
			<SocketContext.Provider>
				<EventsContext.Provider value={{ selectedEvent: event }}>
					<EventPanel />
				</EventsContext.Provider>
			</SocketContext.Provider>
		</DisplaySettingsContext.Provider>
	);
}

test('renders correctly', () => {
	const tree = renderer.create(renderEventPanel(mockEvent)).toJSON();
	expect(tree).toMatchSnapshot();
});

test('should show empty data view if empty', () => {
	render(renderEventPanel(undefined));
	const eventTable = screen.queryByTestId('selected-event');

	expect(eventTable).toBeInTheDocument();
	expect(eventTable).toHaveTextContent(displayData['empty-data-message']);
});

test('should show event home team name', () => {
	render(renderEventPanel(mockEvent));
	const eventTable = screen.queryByTestId('selected-event');

	expect(eventTable).toBeInTheDocument();
	expect(eventTable).toHaveTextContent(mockEvent.competitors[0].name);
});

test('should show event away team name', () => {
	render(renderEventPanel(mockEvent));
	const eventTable = screen.queryByTestId('selected-event');

	expect(eventTable).toBeInTheDocument();
	expect(eventTable).toHaveTextContent(mockEvent.competitors[1].name);
});

test('should show event score', () => {
	const { scores } = mockEvent;
	const { home, away } = scores;
	render(renderEventPanel(mockEvent));
	const eventTable = screen.queryByTestId('selected-event');

	expect(eventTable).toBeInTheDocument();
	expect(eventTable).toHaveTextContent(`${home} - ${away}`);
});

test('should show event start time', () => {
	const { startTime } = mockEvent;

	const renderStartTime = () => {
		const date = new Date(startTime);
		const hour = date.getHours();
		const minutes = date.getMinutes();
		return `${hour}:${minutes}`;
	};

	render(renderEventPanel(mockEvent));
	const eventTable = screen.queryByTestId('selected-event');

	expect(eventTable).toBeInTheDocument();
	expect(eventTable).toHaveTextContent(renderStartTime());
});
