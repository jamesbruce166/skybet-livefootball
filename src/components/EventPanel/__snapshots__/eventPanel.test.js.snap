// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bczRLJ bhMxDI"
  data-testid="selected-event"
>
  <div
    className="sc-ftvSup YVbWV"
  >
    <div
      className="sc-jqUVSM"
    >
      <h2
        className="sc-gsnTZi inIMHQ"
      >
        Chinese Super League
      </h2>
    </div>
    <div
      className="sc-papXJ jNPbEY"
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 lbJwfL sc-kDDrLX ernRby"
        fill="currentColor"
        focusable="false"
        onClick={[Function]}
        onKeyDown={[Function]}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 176a80 80 0 1080 80 80.24 80.24 0 00-80-80zm172.72 80a165.53 165.53 0 01-1.64 22.34l48.69 38.12a11.59 11.59 0 012.63 14.78l-46.06 79.52a11.64 11.64 0 01-14.14 4.93l-57.25-23a176.56 176.56 0 01-38.82 22.67l-8.56 60.78a11.93 11.93 0 01-11.51 9.86h-92.12a12 12 0 01-11.51-9.53l-8.56-60.78A169.3 169.3 0 01151.05 393L93.8 416a11.64 11.64 0 01-14.14-4.92L33.6 331.57a11.59 11.59 0 012.63-14.78l48.69-38.12A174.58 174.58 0 0183.28 256a165.53 165.53 0 011.64-22.34l-48.69-38.12a11.59 11.59 0 01-2.63-14.78l46.06-79.52a11.64 11.64 0 0114.14-4.93l57.25 23a176.56 176.56 0 0138.82-22.67l8.56-60.78A11.93 11.93 0 01209.94 26h92.12a12 12 0 0111.51 9.53l8.56 60.78A169.3 169.3 0 01361 119l57.2-23a11.64 11.64 0 0114.14 4.92l46.06 79.52a11.59 11.59 0 01-2.63 14.78l-48.69 38.12a174.58 174.58 0 011.64 22.66z"
        />
      </svg>
      <div
        className="szh-menu-container"
        onBlur={[Function]}
        onKeyDown={[Function]}
      />
      <button
        className="sc-evZas caOTXG"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="StyledIconBase-ea9ulj-0 lbJwfL sc-iqcoie fzaHUf"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            data-name="Layer 2"
          >
            <g
              data-name="expand"
            >
              <path
                d="M20 5a1 1 0 00-1-1h-5a1 1 0 000 2h2.57l-3.28 3.29a1 1 0 000 1.42 1 1 0 001.42 0L18 7.42V10a1 1 0 001 1 1 1 0 001-1zM10.71 13.29a1 1 0 00-1.42 0L6 16.57V14a1 1 0 00-1-1 1 1 0 00-1 1v5a1 1 0 001 1h5a1 1 0 000-2H7.42l3.29-3.29a1 1 0 000-1.42z"
              />
            </g>
          </g>
        </svg>
      </button>
    </div>
  </div>
  <div
    className="sc-dkzDqf fJxVwt"
  >
    <div
      className="sc-hKMtZM iuTBwl"
    >
      <p
        className="sc-eCYdqJ isgUhi"
      >
        Shanghai Shenhua
      </p>
      <p
        className="sc-eCYdqJ isgUhi"
      >
        0
         - 
        0
      </p>
      <p
        className="sc-eCYdqJ isgUhi"
      >
        Shandong Luneng Taishan
      </p>
    </div>
    <p
      className="sc-jSMfEi iKxlTa"
    >
      Started At
    </p>
    <p
      className="sc-gKXOVf ewIjEI"
    >
      12:35
    </p>
  </div>
</div>
`;
